core/consensus/grandpa/impl/voting_round_impl.cpp

```diff
@@ -922,60 +1363,180 @@ namespace kagome::consensus::grandpa {
     // the round-estimate is the highest block in the chain with head
     // `prevote_ghost` that could have supermajority-commits.
     if (precommits_->getTotalWeight() >= threshold_) {
-      auto block = graph_->findAncestor(
-          BlockInfo{prevote_ghost.block_number, prevote_ghost.block_hash},
-          possible_to_precommit,
-          VoteWeight::precommitComparator);
-      if (block.has_value()) {
-        current_round_state_->best_final_candidate = block.value();
+      auto current_best = prevote_ghost_.has_value() ? prevote_ghost_.value()
+                                                     : last_finalized_block_;
+
+      auto best_final_candidate = graph_->findAncestor(
+          current_best, possible_to_precommit, VoteWeight::precommitComparator);
+      if (best_final_candidate.has_value()) {
+        best_final_candidate_ = best_final_candidate;
+        logger_->debug(
+            "Round #{}: bestFinalCandidate <- best_final_candidate is updated "
+            "by "
+            "precommit_ghost with given equvocations (block #{} hash={})",
+            round_number_,
+            best_final_candidate_->block_number,
+            best_final_candidate_->block_hash);
+      } else {
+        logger_->debug(
+            "Round #{}: bestFinalCandidate <- best_final_candidate is not "
+            "updated by "
+            "precommit_ghost based on prevote_ghost");
       }
     } else {
-      current_round_state_->best_final_candidate =
-          BlockInfo{prevote_ghost.block_number, prevote_ghost.block_hash};
-      return;
+      auto current_best = prevote_ghost_.has_value() ? prevote_ghost_.value()
+                                                     : last_finalized_block_;
+      best_final_candidate_ = current_best;
+      logger_->debug(
+          "Round #{}: bestFinalCandidate <- update best_final_candidate by "
+          "prevote_ghost_ (block #{} hash={})",
+          round_number_,
+          best_final_candidate_->block_number,
+          best_final_candidate_->block_hash);
+      return best_final_candidate_.value();
+    }
+
+    bool maybe_need_handle_completability = not completable_;
+
+    if (best_final_candidate_.has_value()) {
+      if (best_final_candidate_ != bestPrevoteCandidate()) {
+        completable_ = true;
+      } else {
+        auto ghost = graph_->findGhost(best_final_candidate_,
+                                       possible_to_precommit,
+                                       VoteWeight::precommitComparator);
+        if (ghost.has_value()
+            && best_final_candidate_ == bestPrevoteCandidate()) {
+          completable_ = true;
+        } else {
+          logger_->debug(
+              "Round #{}: bestFinalCandidate <- "
+              "precommit ghost (block #{} hash={}) based on best prevote "
```

Comment:
Code doublication.

Reply:
Thanks. I'll try to extract method for that. Or remove if it would needless.